pkg/services/alerting/ticker_test.go

```diff
@@ -1,121 +1,99 @@
-//
-// func TestTickerRetro1Hour(t *testing.T) {
-// offset := time.Duration(10) * time.Second
-// last := time.Unix(0, 0)
-// mock := clock.NewMock()
-// mock.Add(time.Duration(1) * time.Hour)
-// desiredLast := mock.Now().Add(-offset)
-// ticker := NewTicker(last, offset, mock)
-//
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(10)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// }
-//
-// func TestAdvanceWithUpdateOffset(t *testing.T) {
-// offset := time.Duration(10) * time.Second
-// last := time.Unix(0, 0)
-// mock := clock.NewMock()
-// mock.Add(time.Duration(1) * time.Hour)
-// desiredLast := mock.Now().Add(-offset)
-// ticker := NewTicker(last, offset, mock)
-//
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(10)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// // lowering offset should see a few more ticks
-// offset = time.Duration(5) * time.Second
-// ticker.updateOffset(offset)
-// desiredLast = mock.Now().Add(-offset)
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(9)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// // advancing clock should see even more ticks
-// mock.Add(time.Duration(1) * time.Hour)
-// desiredLast = mock.Now().Add(-offset)
-// last = assertAdvanceUntil(ticker, last, desiredLast, offset, time.Duration(8)*time.Millisecond, t)
-// assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-//
-// }
-//
-// func getCase(lastSeconds, offsetSeconds int) (time.Time, time.Duration) {
-// last := time.Unix(int64(lastSeconds), 0)
-// offset := time.Duration(offsetSeconds) * time.Second
-// return last, offset
-// }
-//
-// func TestTickerNoAdvance(t *testing.T) {
-//
-// // it's 00:01:00 now. what are some cases where we don't want the ticker to advance?
-// mock := clock.NewMock()
-// mock.Add(time.Duration(60) * time.Second)
-//
-// type Case struct {
-// 	last   int
-// 	offset int
-// }
-//
-// // note that some cases add up to now, others go into the future
-// cases := []Case{
-// 	{50, 10},
-// 	{50, 30},
-// 	{59, 1},
-// 	{59, 10},
-// 	{59, 30},
-// 	{60, 1},
-// 	{60, 10},
-// 	{60, 30},
-// 	{90, 1},
-// 	{90, 10},
-// 	{90, 30},
-// }
-// for _, c := range cases {
-// 	last, offset := getCase(c.last, c.offset)
-// 	ticker := NewTicker(last, offset, mock)
-// 	assertNoAdvance(ticker, last, time.Duration(500)*time.Millisecond, t)
-// }
-// }
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTicker(t *testing.T) {
+	readChanOrFail := func(t *testing.T, c chan time.Time) time.Time {
+		t.Helper()
+		select {
+		case tick := <-c:
+			return tick
+		default:
+			require.Failf(t, "channel is empty but it should have a tick", "")
+		}
+		return time.Time{}
```

Comment:
It's always a good practice when working with channels on tests, make sure you always set a timeout (that would be shorter than the one used for the whole test run) -- otherwise, there could be cases where the CI just hangs until the global timeout hits.

```golang
	withTimeout, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	t.Cleanup(func() {
		cancel()
	})

	readChanOrFail := func(t *testing.T, ctx context.Context, c chan time.Time) time.Time {
		t.Helper()

		select {
		case <-ctx.Done():
			return time.Time{} // or alternative require.Failf(t, ctx.Err(), "")
		case tick := <-c:
			return tick
		default:
			require.Failf(t, "channel is empty but it should have a tick", "")
		}

		return time.Time{}
	}

```